Fix and improve `split-issue-pr-search-results`

dertieran opened the pull request:
Fixes #3824 

Steps to test as mentioned in the issue:
1. Visit https://github.com/search?q=refined+github&type=Issues
2. Click Pull requests (and wait for load)
3. Click Issues

The problem was that `createUrl` always appended the next `is:<type>` query part.
This is also used by Github for the filter and every previous `is:<type>` is ignored.
But the detection did just check for the first occurence of `is:<type>` so the ui wasn't updated correctly.

I changed the code to first clear all `is:<type>` and then add the new filter.
I also improved the detection if multiple `is:<types>` are in the search.
As a bonus I also added that the language and state search parameters are also copied for each url

fregante commented:
Good catch. The code is a little verbose but I haven’t looked exactly at how it can be improved

fregante commented on source/features/split-issue-pr-search-results.tsx:
```diff
@@ -12,20 +12,51 @@ function cleanLinks(): void {
 
+function getPageSearchQuery(): SearchQuery {
+	const searchParameters = new URLSearchParams(location.search);
+	return new SearchQuery(searchParameters);
```
`getPageSearchQuery` be dropped, SearchQuery accepts location
```suggestion
	return new SearchQuery(location);
```

dertieran replied:
This only works if the `searchQuery` isn't modified otherwise `location.search` is also changed.
So it would not work for `createUrl`, didn't personally like that side effect from `SearchQuery` so explicitly created a new copy 😅

fregante replied:
Indeed, I forgot how it worked. I think the issue here was that we're rebuilding both links from scratch (`createUrl`) starting from `location` instead of just changing the existing `Issues` link. Would you like to try that? This would also avoid the other comment (https://github.com/sindresorhus/refined-github/pull/3837/files#r547508824) since they'll already be in the link

dertieran replied:
Done this :+1:

fregante commented on source/features/split-issue-pr-search-results.tsx:
```diff
@@ -12,20 +12,51 @@ function cleanLinks(): void {
+		if (searchQuery.searchParams.has(name)) {
+			url.searchParams.set(name, searchQuery.searchParams.get(name)!);
+		}
+	}
```
Instead of copying them, you can set the whole search at once.

dertieran replied:
My main thought here was to only copy things that are actually available/used.
E.g. this would avoid copying something like [`?package_type=npm`](https://github.com/search?package_type=npm&q=refined+github&type=RegistryPackages) which has no effect.
But if it is more readable I can also just copy all search params instead.

fregante commented on source/features/split-issue-pr-search-results.tsx:
```diff
@@ -12,20 +12,51 @@ function cleanLinks(): void {
+	const searchQuery = getPageSearchQuery();
+	// Remove all existing is:pr/issue query parts, only the last is used anyways
+	searchQuery.remove('is:pr', 'is:issue');
+	searchQuery.add(`is:${type}`);
```
There’s a replace method that accepts a regex

dertieran replied:
Good idea, didn't think about using a regex 😅

fregante commented on source/features/split-issue-pr-search-results.tsx:
```diff
@@ -47,11 +78,13 @@ function init(): void {
-	if (pageSearchQuery.includes('is:pr')) {
+
+	const is = searchQueryIs(searchQuery);
+	if (is === 'pr') {
```
Instead of this change, perhaps SearchQuery can handle deduplicating opposite flags (issue/pr for this bugfix; later for open/closed, merged/unmerged, etc)

dertieran replied:
Sure why not 👍
I think `SearchQuery` just looked a bit scary to touch with it's side effects 😄

dertieran replied:
When should the query be cleaned?
Right now it is only cleaned when it is `set`, so I added the deduplicate code there.
_(This allows to just use `add` [here](https://github.com/sindresorhus/refined-github/pull/3837#pullrequestreview-557362968))_

Should it also be cleaned on `get` then we could use the `includes` here.
Alternative would also be to add something linke `last(...searchStrings: string[]): string | null {}`.

fregante replied:
As for the reloading, I think it should stop accepting a raw Location, perhaps it should be changed to use the `location.search` string, this way we eliminate the side effect. 

I’m conflicted about where this should go, because it should be a .clean()/.deduplicate() method to make it explicit, but also we don’t want to have to manually call it every time. I think they should be deduplicated:

- in the constructor, for existing values
- in set, for any new values 

So maybe you can implement it as a private method and call it from both places

dertieran replied:
Okay also implemented this now, I just did `this.set(this.get())` to clean the initial query string.
This also means it will update the search for all passed *"sources"*.

fregante reviewed (approved):
Untested but it looks right!

Can you add some tests to ensure that the query is cleaned correctly? Also add some _tricky_ queries to ensure that they're not broken by mistake.

dertieran commented:
@fregante Sure, just not sure what a *"tricky"* query would be :sweat_smile: 
Right now I added a test that uses the query from the issue `?q=refined+github` and the `.add('is:issue')`.
I guess that's fairly simple, would tricky be something with multiple `is:pr/issue`?

Also I noticed that before the passed `URLSearchParams` where just assigned, I changed it that it now creates a new copy, should I change that back?

fregante commented:
I just realized that the regex is acting on the raw string like it did at the beginning of the PR. I’d rather getQueryParts as previously suggested

> SearchQuery was added specifically to avoid having to reparse the query with different code in every feature

dertieran commented:
> I just realized that the regex is acting on the raw string like it did at the beginning of the PR

So you mean something like this?
```js
set(query) {
	const parts = query.split(/\s+/).filter(Boolean);
	this.searchParams.set('q', deduplicate(parts).join(' '));
}
```

dertieran commented:
Okay this now are changes then I initially intended :sweat_smile:  
But I now also removed to `TODO` for supporting spaces in quoted query strings.

fregante commented on test/search-query.ts:
```diff
@@ -4,34 +4,34 @@ import './fixtures/globals';
 
 test('.getQueryParts', t => {
-	const query = new SearchQuery(new URLSearchParams('?q=cool+is%3Aissue'));
+	const query = new SearchQuery('?q=cool+is%3Aissue');
```
I'd rather see a more readable
```suggestion
	const query = new SearchQuery(['q', 'cool is:issue']);
```

dertieran replied:
Sure, so just accept all `URLSearchParams` options?
```js
URLSearchParams(init?: string | string[][] | Record<string, string> | URLSearchParams)
```

*Would be this then*
```js
new SearchQuery([['q', 'cool is:issue']]);
```

But it would also work to just pass the not escaped string if it is just for readability in the tests
```js
new SearchQuery('q=cool is:issue');
```

dertieran replied:
I just changed `SearchQuery` to accept all `URLSearchParams` options.
And then used the `{q: 'query'}` parameter *(thought it was more readable than the entries option)*.

Hope that is okay otherwise I can also change it back or to something else.

fregante commented on source/github-helpers/search-query.ts:
```diff
@@ -1,17 +1,43 @@
+
+// Remove all values from array exept the last occurence of one of the values.
+function deduplicate(array: any[], ...values: any[]): any[] {
+	let indexToKeep = -1;
+	for (const value of values) {
+		const lastIndex = array.lastIndexOf(value);
+		if (lastIndex > indexToKeep) {
+			indexToKeep = lastIndex;
+		}
+	}
+
+	if (indexToKeep === -1) {
+		return array;
+	}
+
+	return array.filter((value, index) => index >= indexToKeep || !values.includes(value));
```
```suggestion
	let alreadyFound = false;
	const newArray = []
	for (const item of array.reverse()) {
		const found = values.includes(item);
		if (!alreadyFound || !found) {
			newArray.unshift(item)
		}
		
		alreadyFound = alreadyFound || found;
	}

	return newArray;
```

dertieran replied:
I think this might be a bit *"slower"* in some edgcases but it's way more readable :+1:
Thanks :heart_eyes:

fregante commented:
Thanks for fixing this!
